### PR TITLE
driver: added hts21 and lps25 support

### PR DESCRIPTION
**Pull request by hoihu**
They are used on the SenseHAT board and are I2c based
temperature/humidity and temperature/pressure sensors from ST Micro

Also included is support for fixed point devices

### CODE REVIEW

**Comment by hoihu** [comment]
I just saw that in #1723 there is a pending issue about where to put the drivers. 

Not sure if it easy to categorize sensors in groups of what they measure, because most sensors measure more than one physical variable. 

Maybe group it in the way they are connected electrically (I2C/SPi/1wire etc)?

File: drivers/i2c/hts21.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+>>> from pyb import I2C
+>>> from hts21 import HTS21
+>>> hts = HTS21(I2C(1, I2C.MASTER), 0x5f)
+>>> hts.read_temp()
+22.06582
+>>> hts.read_humidity()
+57.44328
+"""
+from pyb import I2C
+import pyb
```
Comment (dpgeorge) [inline drivers/i2c/hts21.py]: I don't think you need to import either of these.

File: drivers/i2c/lps25.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+>>> from lps25 import LPS25
+>>> lps = LPS25(I2C(1, I2C.MASTER), 0x5c)
+>>> lps.read_pressure()
+944.4485
+>>> lps.read_temp()
+21.71667
+"""
+
+from pyb import I2C
+import pyb
```
Comment (dpgeorge) [inline drivers/i2c/lps25.py]: Not needed.

**Comment by dpgeorge** [comment]
Great, thank you!

Please make sure it's PEP8 clean: space after comma, no spaces around "=" when used as kw arg.

File: drivers/i2c/hts21.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    
+    def __init__(self, i2c, address, fixed_point = False):
+        """
+        `fixed_point` makes this class use fixed point return values with 2 decimal places 
+        (for hw without floating point support)
+        """
+        self.i2c = i2c
+        self.address = address
+        self.fixed_point = fixed_point
+        self.hts_calib = array.array("h",[0,0,0,0,0,0,0,0,0,0])
```
Comment (dpgeorge) [inline drivers/i2c/hts21.py]: Is it important to use an array?  Could it just be a list of ints?

**Comment by dpgeorge** [comment]
Is it important to support fixed point?

**Comment by hoihu** [comment]
> Is it important to support fixed point?

There were [discussions in the forum about fixed point drivers](http://forum.micropython.org/viewtopic.php?f=11&t=1000&start=20) and I thought it would be a good exercise to implement it.

The WiPy is the only official HW platform that is fixed point I believe. Calculating fixed point output is relative straight forward so it may not be a requirement for every driver and can be added easily by people who actually need it.

A possible compromise: Add a comment in the driver code about the fixed point formula.

**Comment by hoihu** [comment]
Btw, I have also driver code for 8x8 LED Matrix and the keypad of the SenseHAT. Let me know if you're interested in that. (see [my repo](https://github.com/hoihu/projects/tree/master/raspi-hat))

It's a framebuffer approach and maybe it will interfere with your work in #1817 so I thought I'd wait (speed/RAM is not so critical with only 64 pixels ;) )

File: drivers/i2c/hts21.py (hunk 3 above)
Comment (hoihu) [inline drivers/i2c/hts21.py]: I wanted to make sure it a signed 16bit integer list. It's used/filled by the i2c.memread command and `int.from_bytes` (which does not yet support the `signed` kw..). 

I thought using a list of Ints would have forced me to do some expensive conversions (using the struct package for example)

**Comment by hoihu** [comment]
Another consideration is to put all sensor drivers below drivers/sensors. And mass storage drivers like sd card, fram etc under drivers/storage.

Other categories to think of are network and display (or output/input).

**Comment by hoihu** [comment]
I skipped fixed point support and used properties for temperature, pressure and humidity values.

**Comment by pfalcon** [comment]
-1 for usage of properties - they're slow and bloated.

**Comment by hoihu** [comment]
@pfalcon : ok - are there guidelines available for driver implementations "best practices"? Or maybe a "reference driver" as an example?

No problem to revert it back but it would be good to have some simple rules for driver developers (also regarding usage documentation, exception handling etc). Perhaps an entry on the wiki?

This would also help to integrate some the drivers floating around in the forum in the master branch.

**Comment by pfalcon** [comment]
So, one of the esp8266 KS stretch goals is developing consistent suite of drivers for bunch of sensors, that's when conventions and guidelines would be established. As I mentioned, I'm -1 on having properties in the interface of any "core" or "close to core" components of uPy. For one, uPy can be compiled with properties support turned off, and there's good reason for that - supporting properties requires scanning much more of object attribute tree than without them. Smaller devices may want to turn it off, and I personally would like to be able to argue that uPy is as efficient for embedded/IoT tasks as practically possible, without any double-think ("generally we're, but here's an area where we deliberately chose to suck more than needed").

No need to revert anything right away though - the mentioned stretch goal is @dpgeorge's pet project, so worth waiting for his comments.

**Comment by dpgeorge** [comment]
Regardless of their implementation detail, I don't like to use properties.  They should really only be used when you made a class name a simple attribute (eg obj.x), then needed in the future to turn it into a function but must retain backwards compatibility.  Using properties from the outset doesn't seem right.  You want it to be a function so you know that something is going on behind the scenes (in this case an I2C read).

**Comment by hoihu** [comment]
I've removed properties and re-introduced read_xxx methods

**Comment by deshipu** [comment]
> A possible compromise: Add a comment in the driver code about the fixed point formula.

This will still not work on platforms and builds that have the floating point support disabled. I think it's much easier to convert from fixed-point integers to floatingp-point numbers, than the other way around.

**Comment by hoihu** [comment]
@deshipu I did it that way for my initial pull request (it is rebased now, but you can see parts of it in the code comment of damien above)

I'm somewhat torn. While I think it's nice to support the fixed point platform - should we really put in extra code for every driver? And then you'd also need to describe how the return value is scaled (e.g. degres*10 etc...). It would be so much simpler if the return value is directly scaled in the physical unit that you measure.

One aspect for fixed point: the precision of the sensor can be retained easily (from the data registers of the sensors). Using floating point sometimes gives an artificial precision, e.g. 23.45897 deg. and needs to be separatly rounded back to the sensors precision.

**Comment by hoihu** [comment]
Closing this since the sensor and machine API have moved on. 

I have updated the hts221 driver in #3256 and wait for feedback there

**Pull request closed without merging**